<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<HTML>
 <HEAD>
   <TITLE> [Twisted-Python] @inlinecallbacks and AlreadyCalledError in	test cases
   </TITLE>
   <LINK REL="Index" HREF="index.html" >
   <LINK REL="made" HREF="mailto:twisted-python%40twistedmatrix.com?Subject=Re%3A%20%5BTwisted-Python%5D%20%40inlinecallbacks%20and%20AlreadyCalledError%20in%0A%09test%20cases&In-Reply-To=%3CBANLkTikugF0rwUcSZ3O3TE%2BgUoPAB9P%2BBA%40mail.gmail.com%3E">
   <META NAME="robots" CONTENT="index,nofollow">
   <style type="text/css">
       pre {
           white-space: pre-wrap;       /* css-2.1, curent FF, Opera, Safari */
           }
   </style>
   <META http-equiv="Content-Type" content="text/html; charset=utf-8">
   <LINK REL="Previous"  HREF="056300.html">
   <LINK REL="Next"  HREF="056302.html">
 </HEAD>
 <BODY BGCOLOR="#ffffff">
   <H1>[Twisted-Python] @inlinecallbacks and AlreadyCalledError in	test cases</H1>
    <B>Brad Milne</B> 
    <A HREF="mailto:twisted-python%40twistedmatrix.com?Subject=Re%3A%20%5BTwisted-Python%5D%20%40inlinecallbacks%20and%20AlreadyCalledError%20in%0A%09test%20cases&In-Reply-To=%3CBANLkTikugF0rwUcSZ3O3TE%2BgUoPAB9P%2BBA%40mail.gmail.com%3E"
       TITLE="[Twisted-Python] @inlinecallbacks and AlreadyCalledError in	test cases">brad.milne at devx.runthered.com
       </A><BR>
    <I>Mon Apr  4 20:49:50 MDT 2011</I>
    <P><UL>
        <LI>Previous message (by thread): <A HREF="056300.html">[Twisted-Python] @inlinecallbacks and AlreadyCalledError in	test	cases
</A></li>
        <LI>Next message (by thread): <A HREF="056302.html">[Twisted-Python] Gtk2Reactor: what are the reasons for avoiding	GLib functions?
</A></li>
         <LI> <B>Messages sorted by:</B> 
              <a href="date.html#56301">[ date ]</a>
              <a href="thread.html#56301">[ thread ]</a>
              <a href="subject.html#56301">[ subject ]</a>
              <a href="author.html#56301">[ author ]</a>
         </LI>
       </UL>
    <HR>  
<!--beginarticle-->
<PRE>On 5 April 2011 14:21, &lt;<A HREF="https://twistedmatrix.com/cgi-bin/mailman/listinfo/twisted-python">exarkun at twistedmatrix.com</A>&gt; wrote:

&gt;<i> On 01:39 am, <A HREF="https://twistedmatrix.com/cgi-bin/mailman/listinfo/twisted-python">brad.milne at devx.runthered.com</A> wrote:
</I>&gt;<i> &gt;Hi all
</I>&gt;<i> &gt;
</I>&gt;<i> &gt;I have recently started switching to trial.unittest from python's own.
</I>&gt;<i> &gt;The
</I>&gt;<i> &gt;trouble I'm experiencing is when a timeout occurs in my test, it
</I>&gt;<i> &gt;errbacks().
</I>&gt;<i> &gt;Then the @inlineCallbacks decorator sees the error and errbacks(). But
</I>&gt;<i> &gt;then
</I>&gt;<i> &gt;a second @inlineCallback in the chain subsequently sees *that* errback
</I>&gt;<i> &gt;and
</I>&gt;<i> &gt;tries to errback itself. This results in AlreadyCalledError.
</I>&gt;<i> &gt;
</I>&gt;<i> &gt;In the test setup, various services are started. These are tracked and
</I>&gt;<i> &gt;then
</I>&gt;<i> &gt;shutdown again in the teardown. Also, there is some polling that
</I>&gt;<i> &gt;happens as
</I>&gt;<i> &gt;part of the tests (waiting on db activities, for example). These use
</I>&gt;<i> &gt;deferLater calls, which are also tracked and torn down in the teardown.
</I>&gt;<i> &gt;
</I>&gt;<i> &gt;I've tried _suppressAlreadyCalled in various places to no avail.
</I>&gt;<i> &gt;
</I>&gt;<i> &gt;Below is a simple example that shows the problem.
</I>&gt;<i>
</I>&gt;<i> I tried running the sample, it completes without error.  I guess that's
</I>&gt;<i> not what you're seeing?  What version of Twisted do you have?
</I>&gt;<i>
</I>&gt;<i> Jean-Paul
</I>&gt;<i>
</I>&gt;<i> _______________________________________________
</I>&gt;<i> Twisted-Python mailing list
</I>&gt;<i> <A HREF="https://twistedmatrix.com/cgi-bin/mailman/listinfo/twisted-python">Twisted-Python at twistedmatrix.com</A>
</I>&gt;<i> <A HREF="http://twistedmatrix.com/cgi-bin/mailman/listinfo/twisted-python">http://twistedmatrix.com/cgi-bin/mailman/listinfo/twisted-python</A>
</I>&gt;<i>
</I>

Hi Jean-Paul

I'm using 10.2.0 on a mac. Here's my interpreter (using buildout):

#!/opt/local/Library/Frameworks/Python.framework/Versions/2.6/Resources/Python.app/Contents/MacOS/Python
import sys
sys.path[0:0] = [
    '/Users/brad/Development/python/workspace/Python-Integration/src',
    '/Users/brad/Development/python/eggs/nose-0.11.4-py2.6.egg',
    '/Users/brad/Development/python/eggs/MiniMock-1.2.5-py2.6.egg',
    '/Users/brad/Development/python/eggs/suds-0.4-py2.6.egg',
    '/Users/brad/Development/python/eggs/SOAPpy-0.12.4-py2.6.egg',
    '/Users/brad/Development/python/eggs/wstools-0.3-py2.6.egg',
    '/Users/brad/Development/python/eggs/fpconst-0.7.2-py2.6.egg',

 '/Users/brad/Development/python/eggs/Twisted-10.2.0-py2.6-macosx-10.6-x86_64.egg',

 '/Users/brad/Development/python/eggs/zope.interface-3.6.1-py2.6-macosx-10.6-x86_64.egg',

 '/Users/brad/Development/python/eggs/pymongo-1.9-py2.6-macosx-10.6-x86_64.egg',
    '/Users/brad/Development/python/eggs/pytz-2010h-py2.6.egg',
    ]

The python I have at that location (using MacPorts) is 2.6.6.

Following your email I see that 11.0.0 is released so I tried that and got
the same error:
&lt;snip&gt;
[ERROR]
Traceback (most recent call last):
  File
&quot;/Users/brad/Development/python/eggs/Twisted-11.0.0-py2.6-macosx-10.6-x86_64.egg/twisted/internet/defer.py&quot;,
line 1076, in gotResult
    _inlineCallbacks(r, g, deferred)
  File
&quot;/Users/brad/Development/python/eggs/Twisted-11.0.0-py2.6-macosx-10.6-x86_64.egg/twisted/internet/defer.py&quot;,
line 1066, in _inlineCallbacks
    deferred.errback()
  File
&quot;/Users/brad/Development/python/eggs/Twisted-11.0.0-py2.6-macosx-10.6-x86_64.egg/twisted/internet/defer.py&quot;,
line 388, in errback
    self._startRunCallbacks(fail)
  File
&quot;/Users/brad/Development/python/eggs/Twisted-11.0.0-py2.6-macosx-10.6-x86_64.egg/twisted/internet/defer.py&quot;,
line 448, in _startRunCallbacks
    raise AlreadyCalledError
twisted.internet.defer.AlreadyCalledError:
&lt;/snip&gt;

Many thanks
Brad
-------------- next part --------------
An HTML attachment was scrubbed...
URL: &lt;/pipermail/twisted-python/attachments/20110405/b54fe06f/attachment-0001.html&gt;
</PRE>

<!--endarticle-->
    <HR>
    <P><UL>
        <!--threads-->
	<LI>Previous message (by thread): <A HREF="056300.html">[Twisted-Python] @inlinecallbacks and AlreadyCalledError in	test	cases
</A></li>
	<LI>Next message (by thread): <A HREF="056302.html">[Twisted-Python] Gtk2Reactor: what are the reasons for avoiding	GLib functions?
</A></li>
         <LI> <B>Messages sorted by:</B> 
              <a href="date.html#56301">[ date ]</a>
              <a href="thread.html#56301">[ thread ]</a>
              <a href="subject.html#56301">[ subject ]</a>
              <a href="author.html#56301">[ author ]</a>
         </LI>
       </UL>

<hr>
<a href="https://twistedmatrix.com/cgi-bin/mailman/listinfo/twisted-python">More information about the Twisted-Python
mailing list</a><br>
</body></html>
